packages/smooth_app/lib/pages/personalized_ranking_model.dart

```diff
@@ -0,0 +1,104 @@
+import 'package:flutter/material.dart';
+import 'package:openfoodfacts/openfoodfacts.dart';
+import 'package:smooth_app/data_models/product_preferences.dart';
+import 'package:smooth_app/data_models/up_to_date_product_provider.dart';
+import 'package:smooth_app/database/dao_product.dart';
+import 'package:smooth_app/database/local_database.dart';
+import 'package:smooth_app/pages/tmp_matched_product_v2.dart';
+
+enum LoadingStatus {
+  LOADING,
+  LOADED,
+  ERROR,
+}
+
+/// Model that computes the scores and sorts the barcodes accordingly.
+class PersonalizedRankingModel with ChangeNotifier {
+  PersonalizedRankingModel(this.initialBarcodes);
+
+  final List<String> initialBarcodes;
+
+  late LoadingStatus _loadingStatus;
+  String? _loadingError;
+
+  String? get loadingError => _loadingError;
+  LoadingStatus get loadingStatus => _loadingStatus;
+
+  final List<MatchedScoreV2> scores = <MatchedScoreV2>[];
+
+  int? _timestamp;
+
+  /// Refreshes the computations.
+  Future<void> refresh(
+    final LocalDatabase localDatabase,
+    final ProductPreferences productPreferences,
+  ) async {
+    _clear();
+    _asyncLoad(localDatabase, productPreferences);
+  }
+
+  /// Clears the computations.
+  void _clear() {
+    _loadingStatus = LoadingStatus.LOADING;
+    _loadingError = null;
+    scores.clear();
+  }
+
+  /// Returns the loading progress between 0 (min) and 1 (max).
+  double? getLoadingProgress() {
+    if (_loadingStatus != LoadingStatus.LOADING) {
+      return null;
+    }
+    if (initialBarcodes.isEmpty) {
+      return 1;
+    }
+    return scores.length / initialBarcodes.length;
+  }
+
+  /// Removes a barcode from the barcodes and from the scores.
+  void dismiss(final String barcode) {
+    initialBarcodes.remove(barcode);
+    int? index;
+    int i = 0;
+    for (final MatchedScoreV2 score in scores) {
+      if (score.barcode == barcode) {
+        index = i;
+      }
+      i++;
+    }
+    if (index == null) {
+      return;
+    }
+    scores.removeAt(index);
+  }
+
+  /// Computes the scores from the [Product]s extracted one by one from the db.
+  Future<bool> _asyncLoad(
+    final LocalDatabase localDatabase,
+    final ProductPreferences productPreferences,
+  ) async {
+    try {
+      final DaoProduct daoProduct = DaoProduct(localDatabase);
+      for (final String barcode in initialBarcodes) {
+        final Product? product = await daoProduct.get(barcode);
+        if (product == null) {
+          // unlikely, but what shall we do?
```

Comment:
Could happen when the offline feature is in place which would also contain a page to remove certain data sets, so probably fetching again or just skipping as currently